File: H5/H5/System/StringComparer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace System
{
    public abstract class StringComparer : IComparer<string>, IEqualityComparer<string>
    {
        private static readonly StringComparer _ordinal = new OrdinalComparer(false);
        private static readonly StringComparer _ordinalIgnoreCase = new OrdinalComparer(true);

        public static StringComparer Ordinal
        {
            get
            {
                return _ordinal;
            }
        }

        public static StringComparer OrdinalIgnoreCase
        {
            get
            {
                return _ordinalIgnoreCase;
            }
        }

        public int Compare(object x, object y)
        {
            if (x == y) return 0;
            if (x == null) return -1;
            if (y == null) return 1;

            if (x is String sa)
            {
                if (y is String sb)
                {
                    return Compare(sa, sb);
                }
            }

            if (x is IComparable ia)
            {
                return ia.CompareTo(y);
            }

            throw new ArgumentException("At least one object must implement IComparable.");
        }


        public new bool Equals(Object x, Object y)
        {
            if (x == y) return true;
            if (x == null || y == null) return false;

            if (x is String sa)
            {
                if (y is String sb)
                {
                    return Equals(sa, sb);
                }
            }
            return x.Equals(y);
        }

        public int GetHashCode(object obj)
        {
            if (obj == null)
            {
                throw new ArgumentNullException("obj");
            }

            if (obj is string s)
            {
                return GetHashCode(s);
            }
            return obj.GetHashCode();
        }

        public abstract int Compare(String x, String y);
        public abstract bool Equals(String x, String y);
        public abstract int GetHashCode(string obj);
    }


    internal sealed class OrdinalComparer : StringComparer
    {
        private bool _ignoreCase;

        internal OrdinalComparer(bool ignoreCase)
        {
            _ignoreCase = ignoreCase;
        }

        public override int Compare(string x, string y)
        {
            if (Object.ReferenceEquals(x, y)) return 0;
            if (x == null) return -1;
            if (y == null) return 1;

            if (_ignoreCase)
            {
                return String.Compare(x, y, StringComparison.OrdinalIgnoreCase);
            }

            return String.Compare(x, y, false);
        }

        public override bool Equals(string x, string y)
        {
            if (Object.ReferenceEquals(x, y)) return true;
            if (x == null || y == null) return false;

            if (_ignoreCase)
            {
                if (x.Length != y.Length)
                {
                    return false;
                }
                return (String.Compare(x, y, StringComparison.OrdinalIgnoreCase) == 0);
            }
            return x.Equals(y);
        }

        public override int GetHashCode(string obj)
        {
            if (obj == null)
            {
                throw new ArgumentNullException("obj");
            }

            if (_ignoreCase && obj != null)
            {
                return obj.ToLower().GetHashCode();
            }

            return obj.GetHashCode();
        }

        // Equals method for the comparer itself.
        public override bool Equals(Object obj)
        {
            if (!(obj is OrdinalComparer comparer))
            {
                return false;
            }
            return (this._ignoreCase == comparer._ignoreCase);
        }

        public override int GetHashCode()
        {
            string name = "OrdinalComparer";
            int hashCode = name.GetHashCode();
            return _ignoreCase ? (~hashCode) : hashCode;
        }
    }
}

File: H5/H5/Resources/.generated/System/StringComparer.js
    H5.define("System.StringComparer", {
        inherits: [System.Collections.Generic.IComparer$1(System.String),System.Collections.Generic.IEqualityComparer$1(System.String)],
        statics: {
            fields: {
                _ordinal: null,
                _ordinalIgnoreCase: null
            },
            props: {
                Ordinal: {
                    get: function () {
                        return System.StringComparer._ordinal;
                    }
                },
                OrdinalIgnoreCase: {
                    get: function () {
                        return System.StringComparer._ordinalIgnoreCase;
                    }
                }
            },
            ctors: {
                init: function () {
                    this._ordinal = new System.OrdinalComparer(false);
                    this._ordinalIgnoreCase = new System.OrdinalComparer(true);
                }
            }
        },
        methods: {
            Compare: function (x, y) {
                if (H5.referenceEquals(x, y)) {
                    return 0;
                }
                if (x == null) {
                    return -1;
                }
                if (y == null) {
                    return 1;
                }
                var sa;
                if (((sa = H5.as(x, System.String))) != null) {
                    var sb;
                    if (((sb = H5.as(y, System.String))) != null) {
                        return this.compare(sa, sb);
                    }
                }
                var ia;
                if (((ia = H5.as(x, System.IComparable))) != null) {
                    return H5.compare(ia, y);
                }

                throw new System.ArgumentException.$ctor1("At least one object must implement IComparable.");
            },
            Equals: function (x, y) {
                if (H5.referenceEquals(x, y)) {
                    return true;
                }
                if (x == null || y == null) {
                    return false;
                }
                var sa;
                if (((sa = H5.as(x, System.String))) != null) {
                    var sb;
                    if (((sb = H5.as(y, System.String))) != null) {
                        return this.equals2(sa, sb);
                    }
                }
                return H5.equals(x, y);
            },
            GetHashCode: function (obj) {
                if (obj == null) {
                    throw new System.ArgumentNullException.$ctor1("obj");
                }
                var s;
                if (((s = H5.as(obj, System.String))) != null) {
                    return this.getHashCode2(s);
                }
                return H5.getHashCode(obj);
            }
        }
    });
